internal/goofys.go

```diff
@@ -692,20 +692,35 @@ func (fs *Goofys) ForgetInode(
 	op *fuseops.ForgetInodeOp) (err error) {
 
 	fs.mu.Lock()
-
 	inode := fs.getInodeOrDie(op.Inode)
+	fs.mu.Unlock()
+
+	if inode.Parent != nil {
+		inode.Parent.mu.Lock()
+		defer inode.Parent.mu.Unlock()
+	}
+	fs.mu.Lock()
+	defer fs.mu.Unlock()
+
+	// check again to make sure inode still exists now that we have both the fs and parent lock. inode could have been removed by some other operations?
+	inode = fs.inodes[op.Inode]
+	if inode == nil {
+		return
+	}
+
 	stale := inode.DeRef(op.N)
 
 	if stale {
 		delete(fs.inodes, op.Inode)
 		fs.forgotCnt += 1
-		fs.mu.Unlock()
 
 		if inode.Parent != nil {
-			inode.Parent.removeChild(inode)
+			// check again to make sure parent still exists now that we have both the fs and parent lock. Parent could have been removed by some other operations?
+			parent := fs.inodes[inode.Parent.Id]
+			if parent != nil {
+				parent.removeChildUnlocked(inode)
```

Comment:
As shown in the stacktrace, there is at least one race condition where inode is already removed from parent when we try to remove it here. Acquiring both fs and parent locks seems to help.